Add initialization method for the remote workers

JavierCVilla opened the pull request:
Add a new method to the interface offered by PyRDF allowing users to pass a function. This function will be executed on every remote worker as a first step before starting any other computation. 

As a consequence, users can inject and execute custom code on the worker, without being part of the RDataFrame computational graph.

JavierCVilla committed:
Add mechanism to initialize the worker runtime env

- Add broadcastInitialization method to the interface so users can pass
functions to be executed on the worker environment as a first step.

JavierCVilla committed:
Add tests for the initialization method on different backends

etejedor commented on PyRDF/__init__.py:
```diff
@@ -55,3 +55,6 @@ def include(includes_list):
 
     includes.extend(includes_list)
+
+def broadcastInitialization(fun, *args, **kwargs):
```
Now in the right place :)

Since this is the API function, I would prefer to call it something shorter like `initialize`.

etejedor replied:
And we need documentation for this function.

etejedor commented on PyRDF/__init__.py:
```diff
@@ -55,3 +55,6 @@ def include(includes_list):
     includes.extend(includes_list)
+
+def broadcastInitialization(fun, *args, **kwargs):
+    current_backend.registerInitialization(fun, *args, **kwargs)
```
I would not mix snake and camel case. I think the library is using snake case.

JavierCVilla replied:
Indeed, snake case is what [PEP8 recommends](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names).

etejedor commented on PyRDF/backend/Backend.py:
```diff
@@ -53,6 +53,12 @@ def __init__(self, config={}):
+        self.initialization = None
+
+    def registerInitialization(self, fun, *args, **kwargs):
+        import functools
```
I'd prefer to have all imports at the beginning of the module.

etejedor commented on PyRDF/backend/Backend.py:
```diff
@@ -53,6 +53,12 @@ def __init__(self, config={}):
+
+    def registerInitialization(self, fun, *args, **kwargs):
+        import functools
+        self.initialization = functools.partial(fun, *args, **kwargs)
```
So this means the initialization function will be an attribute of the backend. If the user switches backend, they will need to invoke `initialize` again.
I am thinking if we should move this API function to the backend, then.

JavierCVilla replied:
Currently the attribute is assigned to the `Backend` metaclass, hence switching backends should not require users to invoke the `initialize` method again.

The runtime backend (`Local` or `Dist`) will try to run the initialization method on the mapper ([here](https://github.com/JavierCVilla/PyRDF/pull/7/files#diff-e657c12574bba6547cc9fcf27ee09ec7R63)).

etejedor replied:
Ok I see, then good! As a follow up of my comments below, we could then assign a no-op to the initialization function in the constructor of `Backend`.

JavierCVilla replied:
This was not true, since `initialization` was an attribute bound to the instance, not to the class. Last commit fixes this.

etejedor commented on PyRDF/backend/Local.py:
```diff
@@ -60,6 +60,13 @@ def execute(self, generator):
+        # Run initialization method to prepare the worker runtime environment
+        try:
+            self.initialization()
+        except TypeError as e:
```
Won't this be an `AttributeError`?

etejedor commented on PyRDF/backend/Dist.py:
```diff
@@ -278,6 +278,9 @@ def execute(self, generator):
         selected_branches = generator.head_node.get_branches()
 
+        # Avoid having references to self inside the mapper
+        initialization = self.initialization
```
This will raise an `AttributeError` if `self` has not `initialization` attribute. Can we perhaps, in the constructor of every backend, assign `initialization` to an empty function, so that it is a no-op if the user has not specified any initialization function. This way we also avoid the `try` `catches` below.

JavierCVilla replied:
>  Can we perhaps, in the constructor of every backend, assign initialization to an empty function

I really like this approach. I will put `self.initialization = lambda: None` on the `Backend` metaclass so we solve the issue and remove the `try-catch`.

etejedor commented on PyRDF/backend/Backend.py:
```diff
@@ -53,6 +53,12 @@ def __init__(self, config={}):
+        # Users can set a function to initialize the environment
+        self.initialization = None
+
+    def registerInitialization(self, fun, *args, **kwargs):
```
https://memegenerator.net/img/instances/51744399/document-everything.jpg

JavierCVilla committed:
Address comments

- Use a simpler name in the API (initialize)
- Use snake case rather than camelCase
- initialization method is no-op function by default
- Remove try-catches

Modify the mechanism to share the same initialization method among
different backends, so users do not need to define it again after
switching backend.

- Use class methods so the function is bound to the Backend class rather
than to the different backend instances.

JavierCVilla commented on PyRDF/__init__.py:
```diff
@@ -55,3 +56,25 @@ def include(includes_list):
+        Keyword arguments used to execute the function.
+
+    """
+    Backend.register_initialization(fun, *args, **kwargs)
```
`initialize` may have some limitations, since not every function can be pickled. If the function passed by the user is not a pure function, then pickle will complain at runtime, making it harder for the user to understand the issue. We can add a couple of checks prior the call to `register_initialization` where we try to pickle the user function and continue only if it succeeds.

JavierCVilla replied:
@etejedor What do you think about this?

JavierCVilla replied:
Indeed, this might be part of a more general series of checks in the local machine before going to a distributed environment (added as an issue #12).

JavierCVilla committed:
[NFC] Remove unnecessary import and add correct tab

etejedor commented on PyRDF/backend/Backend.py:
```diff
@@ -54,6 +55,40 @@ def __init__(self, config={}):
         """
         self.config = config
 
+    @classmethod
```
I would define this just as a class attribute before `__init__`:
```suggestion
    initialization = lambda: None
```
and add the documentation about the attribute to the docstring of class `Backend`. BTW I just realized that the `config` attribute is documented both in the class and for the instance, and it should be documented only in the instance (there is no class config).

JavierCVilla replied:
> I would define this just as a class attribute before `__init__`

I tried that, but then I got the following error:

```
TypeError: unbound method <lambda>() must be called with Backend instance as first argument (got nothing instead)
```

It could be a class attribute, but as soon as we call it, it will fail and would require us to run something link:

```
b = Backend()
b.initialization()
```

etejedor replied:
`initialization = lambda cls:None` ?

JavierCVilla replied:
That does not work either, it tries to run:

```
Backend().initialization(cls)
```

which would end up passing the `cls` parameter to the user's `initialize` function.

JavierCVilla committed:
Define initialization as a class attribute
